pkg/traefik/traefik.go

```diff
@@ -0,0 +1,310 @@
+package traefik
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/celestiaorg/knuu/pkg/k8s"
+	"github.com/celestiaorg/knuu/pkg/names"
+	"github.com/sirupsen/logrus"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/utils/ptr"
+)
+
+const (
+	traefikServiceName = "traefik"
+	Port               = 80
+	PortSecure         = 443
+	deploymentName     = "traefik-deployment"
+	roleName           = "traefik-role"
+	containerName      = "traefik"
+	image              = "traefik:v3.0"
+	appLabel           = "app"
+	appLabelValue      = "traefik"
+	replicas           = 2
+	waitRetry          = 5 * time.Second
+)
+
+type Traefik struct {
+	K8s      *k8s.Client
+	endpoint string
+}
+
+func (t *Traefik) Deploy(ctx context.Context) error {
+	if t.K8s == nil {
+		return ErrTraefikClientNotInitialized
+	}
+
+	// Create a dedicated service account for Traefik
+	serviceAccountName, err := names.NewRandomK8("traefik-service-account")
+	if err != nil {
+		return err
+	}
+	if err := t.K8s.CreateServiceAccount(ctx, serviceAccountName, nil); err != nil {
+		return ErrFailedToCreateServiceAccount.Wrap(err)
+	}
+
+	clusterRoleName, err := names.NewRandomK8(roleName)
+	if err != nil {
+		return err
+	}
+
+	// Define and create a ClusterRole for Traefik
+	err = t.K8s.CreateClusterRole(ctx, clusterRoleName, nil, []rbacv1.PolicyRule{
+		{
+			APIGroups: []string{""}, // Core group
+			Resources: []string{"pods", "endpoints", "secrets", "services"},
+			Verbs:     []string{"get", "list", "watch"},
+		},
+		{
+			APIGroups: []string{"traefik.io"}, // Traefik specific resources
+			Resources: []string{
+				"ingressroutes", "middlewares", "tlsstores", "serverstransporttcps",
+				"traefikservices", "ingressrouteudps", "middlewaretcps", "tlsoptions",
+				"serverstransports", "ingressroutetcps",
+			},
+			Verbs: []string{"get", "list", "watch"},
+		},
+		{
+			APIGroups: []string{"networking.k8s.io"}, // Networking resources
+			Resources: []string{"ingresses", "ingressclasses"},
+			Verbs:     []string{"get", "list", "watch"},
+		},
+	})
+
+	if err != nil {
+		return ErrTraefikRoleCreationFailed.Wrap(err)
+	}
+
+	if err := t.K8s.CreateClusterRoleBinding(ctx, clusterRoleName, nil, clusterRoleName, serviceAccountName); err != nil {
+		return ErrTraefikRoleBindingCreationFailed.Wrap(err)
+	}
+
+	// Create the Traefik deployment using the service account
+	traefikDeployment := &appsv1.Deployment{
```

Comment:
Instead of manually deploying the Deployment like that, would it be possible to use a knuu Instance?